src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -90,70 +106,68 @@ struct TensorContract
-    tensor_index_in[I] = tensor_index[I - 2];
-    fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
-  }
-
-  template <size_t I, size_t Rank, Requires<(I == Rank - 1)> = nullptr>
-  SPECTRE_ALWAYS_INLINE void fill_contracting_tensor_index(
-      std::array<int, Rank>& tensor_index_in,
-      const std::array<int, num_tensor_indices>& tensor_index) const {
-    tensor_index_in[I] = tensor_index[I - 2];
+      std::array<size_t, Rank>& tensor_index_in,
+      const std::array<size_t, num_tensor_indices>& tensor_index) const {
+    if constexpr (I < Index1::value) {
+      tensor_index_in[I] = tensor_index[I];
+      fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
+    } else if constexpr (I == Index1::value) {
+      // 10000 is for the slot that will be set later. Easy to debug.
+      tensor_index_in[I] = 10000;
+      fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
+    } else if constexpr (I > Index1::value and I <= Index2::value and
+                         I < Rank - 1) {
+      // tensor_index is Rank - 2 since it shouldn't be called for Rank 2 case
+      // 20000 is for the slot that will be set later. Easy to debug.
+      tensor_index_in[I] = I == Index2::value ? 20000 : tensor_index[I - 1];
+      fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
+    } else if constexpr (I > Index2::value and I < Rank - 1) {
+      // Left as Rank - 2 since it should never be called for the Rank 2 case
+      tensor_index_in[I] = tensor_index[I - 2];
+      fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
+    } else if constexpr (I == Index2::value) {
+      tensor_index_in[I] = 20000;
+    } else {
+      tensor_index_in[I] = tensor_index[I - 2];
+    }
   }
 
   template <typename... LhsIndices, typename U>
-  SPECTRE_ALWAYS_INLINE type
-  get(const std::array<U, num_tensor_indices>& new_tensor_index) const {
+  SPECTRE_ALWAYS_INLINE decltype(auto) get(
+      const std::array<U, num_tensor_indices>& new_tensor_index) const {
     // new_tensor_index is the one with _fewer_ components, ie post-contraction
-    std::array<int, tmpl::size<Symm>::value> tensor_index;
+    std::array<size_t, tmpl::size<Symm>::value> tensor_index{};
     // Manually unrolled for loops to compute the tensor_index from the
     // new_tensor_index
     fill_contracting_tensor_index<0>(tensor_index, new_tensor_index);
     return detail::ComputeContractionImpl<CI1::dim - 1, Index1, Index2>::
         template apply<LhsIndices...>(tensor_index, t_);
   }
 
+  template <typename LhsStructure, typename... LhsIndices>
+  SPECTRE_ALWAYS_INLINE decltype(auto) get(
+      const size_t lhs_storage_index) const {
+    const std::array<size_t, num_tensor_indices>& new_tensor_index =
+        LhsStructure::template get_canonical_tensor_index<num_tensor_indices>(
+            lhs_storage_index);
+    return get<LhsStructure, LhsIndices...>(new_tensor_index);
+  }
```

Comment:
This very basic implementation for a storage index `get` is just to have something minimally working, for now, and not necessarily something that is great. Look out for an improvement on this in a future PR!